Update defaults and use of Internal system

bobjacobsen opened the pull request:
This updates the creation and management of the automatically-defined Internal system and how it interacts with the default mechanism.

- The automatically-created Internal system still contains Lights, Reporters, Sensors, Turnouts but no longer contains Programmers, and Consist and Throttle managers. 
- If you want internals of those, add another system via the usual connections screen
- The automatic setting and changing of defaults while adding and removing systems has been simplified.
- Several changes have been made to how the DecoderPro startup wizard interacts with defaults
- Includes some updates to NCE system internals, particularly support for the various USB modes

Contains the updates in #6160. Expected to fix #6060

bobjacobsen commented:
Restart Travis graphical after:
```
INFO  - Starting with profile Sprog_Simulator.3d40b3b8 [Time-limited test] apps.Apps.<init>()
Exception in thread "init prefs" java.util.ServiceConfigurationError: jmri.swing.PreferencesPanel: Provider jmri.web.server.WebServerPreferencesPanel could not be instantiated
	at java.util.ServiceLoader.fail(ServiceLoader.java:232)
	at java.util.ServiceLoader.access$100(ServiceLoader.java:185)
	at java.util.ServiceLoader$LazyIterator.nextService(ServiceLoader.java:384)
	at java.util.ServiceLoader$LazyIterator.access$700(ServiceLoader.java:323)
	at java.util.ServiceLoader$LazyIterator$2.run(ServiceLoader.java:407)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.util.ServiceLoader$LazyIterator.next(ServiceLoader.java:409)
	at java.util.ServiceLoader$1.next(ServiceLoader.java:480)
	at apps.gui3.TabbedPreferences.init(TabbedPreferences.java:184)
	at apps.Apps.lambda$null$2(Apps.java:387)
	at java.util.Optional.ifPresent(Optional.java:159)
	at apps.Apps.lambda$new$3(Apps.java:386)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NullPointerException: Required nonnull default for apps.StartupActionsManager does not exist.
	at jmri.InstanceManager.getDefault(InstanceManager.java:203)
	at jmri.web.server.WebServerPreferencesPanel.isStartupAction(WebServerPreferencesPanel.java:187)
	at jmri.web.server.WebServerPreferencesPanel.initComponents(WebServerPreferencesPanel.java:73)
	at jmri.web.server.WebServerPreferencesPanel.<init>(WebServerPreferencesPanel.java:45)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.lang.Class.newInstance(Class.java:442)
	at java.util.ServiceLoader$LazyIterator.nextService(ServiceLoader.java:380)
	... 10 more
INFO  - Starting with profile tmcc.3f0b414a [Time-limited test] apps.Apps.<init>()
INFO  - Starting with profile LocoNet_Simulator.3eac0cdc [Time-limited test] apps.Apps.<init>()
INFO  - Starting with profile EasyDcc_Simulator.3f033c2b [Time-limited test] apps.Apps.<init>()
INFO  - Starting with profile Prevent_Init_Loop.3f01644a [Time-limited test] apps.Apps.<init>()
INFO  - Starting with profile tmcc.3f0b414a [Time-limited test] apps.Apps.<init>()
INFO  - Starting with profile LocoNet_Simulator.3eac0cdc [Time-limited test] apps.Apps.<init>()
INFO  - Starting with profile EasyDcc_Simulator.3f033c2b [Time-limited test] apps.Apps.<init>()
[ERROR] Errors: 
[ERROR]   ConnectionConfigXmlTest>AbstractSerialConnectionConfigXmlTestBase.loadTest:40 » ConcurrentModification
[INFO] 
```

dheap commented:
I've been tied up with our Convention all weekend but have now been doing some testing of various scenarios with NCE Power Pro, NCE Power Cab and SPROG systems.

I've found some anomalies, but will have to repeat and document each one. I won't get them all done in what's left of today.

bobjacobsen commented:
@dheap Thank you!  Please take your time, I appreciate the effort it takes to track down anomalies.

There were a couple places where a SystemConnectionMemo object was being created and registered (which is necessary for the defaults to be selected), but it's contents where being updated later (i.e. causing it to change whether it provided a CommandStation or not).  I might have missed some of those.

dheap commented:
Scenario 1: create NCE SB5 using manual-internal-11 with wizard
![screen shot 2018-11-19 at 4 20 15 pm](https://user-images.githubusercontent.com/13841457/48687817-3147eb80-ec17-11e8-98f9-4972f5822ffc.png)
(use any port if no SB5 hardware)
When wizard completed: 
![screen shot 2018-11-19 at 4 23 01 pm](https://user-images.githubusercontent.com/13841457/48687903-b29f7e00-ec17-11e8-83aa-ace94c0c37e4.png)
**Not what we want: should have Ops Mode programmer available.**
[create_NCE_SB5_using_manualinternal11_with_wizard.jmri.zip](https://github.com/JMRI/JMRI/files/2593889/create_NCE_SB5_using_manualinternal11_with_wizard.jmri.zip)
Quit JMRI
[create_NCE_SB5_using_manualinternal11_with_wizard.jmri 2.zip](https://github.com/JMRI/JMRI/files/2593893/create_NCE_SB5_using_manualinternal11_with_wizrd.jmri.2.zip)
and restart:
![screen shot 2018-11-19 at 4 33 34 pm](https://user-images.githubusercontent.com/13841457/48688120-e62ed800-ec18-11e8-9ec1-d883c902960a.png)
That looks good (Service Programmer is not available, but Ops Programmer is, so are throttles etc. when you create a roster entry:
![screen shot 2018-11-19 at 4 38 42 pm](https://user-images.githubusercontent.com/13841457/48688299-b16f5080-ec19-11e8-8612-e2394eeb6caf.png)

So the issue is present after Wizard creation until next startup.

dheap commented:
Scenario 2: create SPROG using manual-internal-11 with wizard
![screen shot 2018-11-19 at 4 47 48 pm](https://user-images.githubusercontent.com/13841457/48688578-e62fd780-ec1a-11e8-90ff-79e8fcba2ce6.png)
(use any port if no SPROG hardware)
When wizard completed:
![screen shot 2018-11-19 at 4 49 30 pm](https://user-images.githubusercontent.com/13841457/48688634-24c59200-ec1b-11e8-972d-0db4879124fd.png)
That seems OK. Restart JMRI and all looks well.
![screen shot 2018-11-19 at 4 51 29 pm](https://user-images.githubusercontent.com/13841457/48688784-b9c88b00-ec1b-11e8-916b-f93965c0e33b.png)
No problems with that scenario, but I'll post it anyway...

dheap commented:
Scenario 3: create SPROG Command Station using manual-internal-11 with wizard
![screen shot 2018-11-19 at 4 59 45 pm](https://user-images.githubusercontent.com/13841457/48688977-8afee480-ec1c-11e8-8080-99c317480c36.png)
(use any port if no SPROG hardware)
When wizard completed:
![screen shot 2018-11-19 at 5 00 14 pm](https://user-images.githubusercontent.com/13841457/48689038-dc0ed880-ec1c-11e8-8ce5-79351b9c8606.png)
That seems OK. Restart JMRI and all looks well.
![screen shot 2018-11-19 at 5 03 26 pm](https://user-images.githubusercontent.com/13841457/48689103-15dfdf00-ec1d-11e8-98a3-18ef71ab1539.png)
[create_SPROG_Command_Station_using_manualinternal11_with_wizard.jmri.zip](https://github.com/JMRI/JMRI/files/2593971/create_SPROG_Command_Station_using_manualinternal11_with_wizard.jmri.zip)
No problems with that scenario, but I'll post it anyway...

dheap commented:
Looking into this now... BRB

dheap commented:
@bobjacobsen I'll call it an evening now and leave you to look at differences in wizard results between SB5 and SPROG Command Station.

I'll be back (hopefully tomorrow) with some V4.12 -> [manual-internal-11](https://github.com/bobjacobsen/JMRI/tree/manual-internal-11) migration anomalies.

dheap commented:
Created 5 Hardware Profiles with V4.12 (as would be done by user who only used Production Releases)
[Archive.zip](https://github.com/JMRI/JMRI/files/2597457/Archive.zip)
Disconnected all hardware (simulating user who forgets to plug in hardware at some stage).

Opened each profile in turn and then quit immediately accepting all dialog box defaults (simulating user who realises hardware unplugged and bails out).
[Archive 2.zip](https://github.com/JMRI/JMRI/files/2597463/Archive.2.zip)

Then opened each Hardware Profile with [manual-internal-11](https://github.com/bobjacobsen/JMRI/tree/manual-internal-11).

**Results with Power Cab:**
![screen shot 2018-11-20 at 10 14 28 am](https://user-images.githubusercontent.com/13841457/48740833-1b850580-ecad-11e8-9f80-79dd718655ba.png)
Not good. Nothing available and no warnings.

**Results with Power Pro:**
![screen shot 2018-11-20 at 10 29 37 am](https://user-images.githubusercontent.com/13841457/48741443-3ce6f100-ecaf-11e8-895f-8c25da43e52a.png)
Works, but Preferences look ugly.

**Results with SPROG Command Station:**
![screen shot 2018-11-20 at 10 27 44 am](https://user-images.githubusercontent.com/13841457/48741366-f5606500-ecae-11e8-9db8-7dfb9cde6337.png)
Works, but Preferences look ugly.

**Results with SPROG:**
![screen shot 2018-11-20 at 10 31 30 am](https://user-images.githubusercontent.com/13841457/48741564-b4b51b80-ecaf-11e8-83c4-aa90b8c379a6.png)
![screen shot 2018-11-20 at 10 31 59 am](https://user-images.githubusercontent.com/13841457/48741623-e928d780-ecaf-11e8-97cb-933e60feed1b.png)
Warning, Preferences look ugly, but works despite warning.
[Archive 3.zip](https://github.com/JMRI/JMRI/files/2597453/Archive.3.zip)

SB5 being redone due to operator error 😕

dheap commented:
Created SB5 Profile with V4.12 (as would be done by user who only used Production Releases)

Disconnected all hardware (simulating user who forgets to plug in hardware at some stage).

Opened SB5 profile and then quit immediately accepting all dialog box defaults (simulating user who realises hardware unplugged and bails out).
[V4.12_SB5.zip](https://github.com/JMRI/JMRI/files/2597627/V4.12_SB5.zip)

Then opened SB5 Profile with [manual-internal-11](https://github.com/bobjacobsen/JMRI/tree/manual-internal-11).

**Results with SB5:**
![screen shot 2018-11-20 at 11 59 00 am](https://user-images.githubusercontent.com/13841457/48744693-5cd0e180-ecbc-11e8-878f-b24061a98ed4.png)
All looks good.
[V4.12_SB5 2.zip](https://github.com/JMRI/JMRI/files/2597639/V4.12_SB5.2.zip)

bobjacobsen commented:
Thank you for all the detailed help on this!

The NCE issues are due to the NCE start-up process not calling the configure() method at the right time. Finding and fixing something that’s _not_ happening can be a bit tricky. My first fix broke other things. I’m still working on it and should have a (hopefully correct) fix in a day or two.

bobjacobsen commented:
@dheap I could use your help on how some of the NCE code is supposed to work.  During initialization of the programmers, it encounters this code:
```
         if (getNceUsbSystem() != NceTrafficController.USB_SYSTEM_NONE) {
             if (getNceUsbSystem() != NceTrafficController.USB_SYSTEM_POWERHOUSE) {
                 // don't set a service mode programmer
             }
```

That means that _any_ USB-connected NCE system doesn't get a service mode programmer.  As a side-effect, if it goes through this code _before_ another part of the initialization, it also doesn't get a addressed programmer.  (That order-dependent initialization is what you're seeing in the wizard vs normal startup case you found)

So two questions:
- Is there a reason that POWERHOUSE gets special treatment here? Is it somehow different?  The effect of the case is to treat it the same _except_ for a message.  Is there some other difference I should code in?
- Are there other NCE USB hardware setups that have addressed programmers besides the SB5 case? I.e. PowerCab, SB3, Twin?

Thanks!

dheap commented:
The PowerHouse can be connected by two possible methods:
1. The DB9 Serial Port, in which case it supports both service and addressed programmers.
2. The NCE USB board, in which case it has _no_ programmers.

All other systems need the NCE USB board:

1. The Power Cab and DCC Twin support both service and addressed programmers.
2. The SB3/3a and SB5 support addressed programmer but _no_ service programmer.

Furthermore:
- For the USB-only systems, various other features are enabled/disabled depending on both the system type and the NCE USB board version (V6.x.x versus V7.x.x).
- The NCE USB board does not support any raw packet commands on any system. For some packet types we are able to intercept, extract the data and convert the raw packet command to an NCE binary command. See `NceTrafficController.sendPacket()` (an incomplete implementation of the `CommandStation` interface).

See also [USB Adapter Limitation Summary](http://jmri.org/help/en/html/hardware/nce/NCE.shtml#usblimitations) (recently updated).

KenC57 commented:
My memory of the code was intending to say:

If the system is using the NCE USB, then if the command station is a
PowerHouse (ie not a USB type command station), you don't get a programmer
(since the code in the PowerHouse doesn't support the commands over the cab
bus) it only does it via the programming commands on the serial port. That
set up you barely get throttle support. There is no space in the PowerHouse
firmware to know those commands.

-ken c

bobjacobsen commented:
> If the system is using the NCE USB, then if the command station is a
> PowerHouse (ie not a USB type command station), you don't get a programmer
> (since the code in the PowerHouse doesn't support the commands over the cab
> bus) it only does it via the programming commands on the serial port. That
> set up you barely get throttle support. There is no space in the PowerHouse
> firmware to know those commands.


What the code actually says: If using a NCE USB, you don’t get a (global) programmer.  If it’s a USB connection to a Powerhouse, you still don’t get a global programmer (see prior sentence) but you also log a debug message.

bobjacobsen commented:
Thanks!  Will update to have this logic.

> The PowerHouse can be connected by two possible methods:
> 
> 	• The DB9 Serial Port, in which case it supports both service and addressed programmers.
> 	• The NCE USB board, in which case it has no programmers.
> All other systems need the NCE USB board:
> 
> 	• The Power Cab and DCC Twin support both service and addressed programmers.
> 	• The SB3/3a and SB5 support addressed programmer but no service programmer.
>

bobjacobsen commented:
> But that code excerpt does look a bit odd. Where's it from?

NceSystemConnectionMemo#configureManagers line 210

https://github.com/JMRI/JMRI/blob/ebd38bdec4279067600b747e5a61e1f7f6cbe116/java/src/jmri/jmrix/nce/NceSystemConnectionMemo.java#L210

bobjacobsen commented:
> 
> NceSystemConnectionMemo#configureManagers line 210

Sorry, that link was to after the update.  Here’s the before:

https://github.com/JMRI/JMRI/blob/d708a33b942b4f24a00f5b5446bdf37dcc4fada5/java/src/jmri/jmrix/nce/NceSystemConnectionMemo.java#L211

bobjacobsen commented:
@dheap, there's also logic in NceProgrammer that seems to be making some (odd) choices around programming modes (in Service mode).  Can you confirm that 

- the ProCab and Twin can do Register, Paged and Direct mode regardless of the getCommandOptions() value?
- that for non-USB connections, Direct is only available for `getCommandOptions() >= NceTrafficController.OPTION_2006`

Thanks again!

dheap commented:
> the ProCab and Twin can do Register, Paged and Direct mode regardless of the getCommandOptions() value?

NCE terminology is confusing:

- PowerHouse is the obsolete name for the Power Pro command station.
- ProCab is a throttle for use on the cab bus of any NCE system.
- Power Cab is an all-in-one DCC system. It is effectively a ProCab with a command station and booster included. Standalone it functions as a full DCC system, with support for external cab bus devices. If plugged into the cab bus of another DCC system it detects existing cab bus traffic and reverts to being a "dumb" ProCab. It supports all programming modes but has a single shared main/program track output.
- The SB3/3a/5 so-called "Smart Boosters" are actually a command station and booster combined, with no program track circuitry.
- The DCC Twin is an entry-level system effectively consisting of a command station, booster and programmer (very like the Power Cab) incorporated into a box with two front panel centre-off simplified throttles. Its appearance is very reminiscent of the Hammant & Morgan Duette and is probably targeted at that market. Like the Power Cab, it can support external cab bus devices or run in slave mode as two simplified throttles on another DCC system,

So I'm guessing your question should have been 
> the _Power Cab_ and Twin can do Register, Paged and Direct mode regardless of the getCommandOptions() value?

In this case the answer is Yes.

> that for non-USB connections, Direct is only available for getCommandOptions() >= NceTrafficController.OPTION_2006

This is quite possible. Power Pro command station firmware prior to 2006 used an ASCII communications protocol (cf. SPROG) versus the Binary Byte protocol for all current systems. I've dug up an old manual and it appears that only two modes were supported:

1. Read/Write Register.
2. Read/Program CV.

I'm guessing (2) would be Paged Mode.

bobjacobsen commented:
Thanks!

I've done some refactoring and simplification of the NCE "who's got which programmer" logic.  Two remaining things:
- In NceProgramManager, while sorting out whether there's a Global (Service Mode) programmer, there's this logic:
```
        if (tc != null && (tc.getUsbSystem() != NceTrafficController.USB_SYSTEM_NONE)) {
            if ((tc.getCmdGroups() & NceTrafficController.CMDS_PROGTRACK) == 0) {
                log.trace("checkGlobalProgrammerAvailable return false");
                return false;
            } else {
                log.trace("checkGlobalProgrammerAvailable return false");
                return true;
            }
        } else { /* stuff for serial */}
```

That doesn't look at the USB type value, but rather at `getCmdGroups()` in the TrafficController.  Is that actually the right thing to do, or should it be decoding the `getUsbSystem()` value?

- One of the USB codes is `USB_SYSTEM_POWERHOUSE`, which (from the above comment) seems to be obsolete terminology for "PowerHouse Pro via USB"  (or maybe "PowerPro via USB"?).  Happy to change it.  What's a better name?

Thanks again

dheap commented:
@bobjacobsen I'm working on and testing some suggested changes/simplifications to `isAddressedModePossible()`and `checkGlobalProgrammerAvailable()`.

Will be back when done...

dheap commented:
@bobjacobsen, @KenC57 

- See my https://github.com/bobjacobsen/JMRI/pull/8 for suggested cleanup changes to `isAddressedModePossible()` and `checkGlobalProgrammerAvailable()`.
- I'm inclined to refactor `USB_SYSTEM_POWERHOUSE` to `USB_SYSTEM_POWERPRO`. I think that's sufficiently descriptive of "PowerPro via USB" and consistent in format with the other `USB_SYSTEM_xxxxx` names. I can quickly refactor using NetBeans if we agree on a name.

bobjacobsen commented:
Thank you!  The POWERPRO name sounds right to me then. Please go ahead and refractor on the current branch, then I’ll merge that PR.

Thanks again

dheap commented:
Done, see https://github.com/bobjacobsen/JMRI/pull/8.

bobjacobsen commented:
#bobjacobsen#8 merged into here, CI running

KenC57 commented:
To the best of my memory, the intent of the getUsbSystem() was to know if it
was using the NCE-USB (meaning over the cab bus) and which type of command
station it was. So USB_SYSTEM_POWERHOUSE is a holdover from the older name
version. Using PowerPro would be more up to date, but key is any command
station that has a serial port on it (that does all the fancy stuff like
programming or memory commands) is what it matches. Focus here was which
command station on the USB.

The getCmdGroups() came a little bit later to make it simpler to know that
the current connection (that is using the NCE-USB) can do which features.
Focus here was which features were supported for this command station (but
only over the USB I think, not sure).

So if you see you are using the NCE-USB, then the getCmdGroups() is the
right thing to use to figure out the rest. I don't know if the code that
configures (setCmdGroups()??) it knows the right things to set for the old
serial interface command stations. That's been enough years I'm not sure
(and having been in the St Lawrence River looking at flooded remains of an
old train station an creamery all day in 41F water may also impact current
memory). I may or may not have made the getCmdGroups() right for all
interfacing types.
-ken c

dheap commented:
Given that `USB_SYSTEM_NONE` covers all  the non-USB cases (currently Serial Port and Simulator), use of the USB system type as the sole selector for availability of Ops and/or Service Programmers in `isAddressedModePossible()` and `checkGlobalProgrammerAvailable()` (rather than a partial selector) seems adequate to me.

I've tested both Programmer availability and programming commands used for each NCE system type (excluding the older Power Pro firmware versions) using this branch and all was correct (to the extent of my observations).

I can't see that we've broken anything and the under-consideration code (which only affects programmer availability, not command sets used) seems more logical (and easier to follow/debug/change) to me.

I haven't changed the functionality of `getCmdGroups()` or `setCmdGroups()` as programming command set selectors.

bobjacobsen commented:
@dheap I did some tests with the Wizard, but without hardware.  In all the tests, the wizard left the right programmers in place, which then turned offline after a short interval.  

But additional testing would be appreciated.  I'd like to merge this pretty soon, so we can get some wider distribution to flush out any remaining issues.

dheap commented:
I've been doing some tests with new profiles and Wizard and hit a snag:

With NCE DCC Twin, while Programming Track is available, opening Single CV Programmer or Identify or Read Type from Decoder, Read Sheets etc. are not available:
`2018-11-27 16:56:51,159 nce.NceMessage                        ERROR - attempt to send unsupported binary command READ_DIR_CV_CMD to NCE USB [AWT-EventQueue-0]
`

I've checked out a branch from [manual-internal-11](https://github.com/bobjacobsen/JMRI/tree/manual-internal-11)at 835e6ae and the problem existed then, predating my changes.

May not have time to do much more tonight...

dheap commented:
Maybe `jmri.jmrix.nce.usbdriver.UsbDriverAdapter.configure()` is not running early enough in the initialisation process, or not having the desired effect...

Or maybe `jmri.jmrix.nce.NceTrafficController.getCmdGroups()` is not being consulted...

bobjacobsen commented:
> On Nov 26, 2018, at 10:30 PM, Dave Heap <notifications@github.com> wrote:
> 
> Maybe jmri.jmrix.nce.usbdriver.UsbDriverAdapter.configure() is not running early enough in the initialisation process, or not having the desired effect...
> 
> Or maybe jmri.jmrix.nce.NceTrafficController.getCmdGroups() is not being consulted…

No, I think it’s that inconsistency in the code.  Before this work, e.g. in 4.13.5, Direct Mode wasn’t available to the Twin.  It is now.

Maybe it shouldn’t be, in which case I can turn it off. Or if it should be, we need to fix a couple more places to know about it.

dheap commented:
The Twin can do them. It can do the same  commands as the PowerCab V1.65x with USB V7.x.x.

That code is incorrect. See 'jmri.jmrix.nce.usbdriver.UsbDriverAdapter.configure()', which is/should be the definitive definer of available commands, retrievable by `NceTrafficController.getCmdGroups()`, as Ken was saying earlier.

Seems `NceProgrammer.getSupportedModes()` is maintaining a duplicate (incorrect) list..

Family duties call...

> There's code in the getReadDirectCV and getWriteDirectCV methods in NceMessage that says those methods are invalid for the Twin.
> 
> * It could be that's wrong, so that part of the check can just be removed.
> * Or, if that's right, I can change NceProgrammer getSupportedModes to say that the Twin can't do it.
> 
> Not clear to me which is the right choice.

dheap commented:
>'jmri.jmrix.nce.usbdriver.UsbDriverAdapter.configure()', which is/should be the definitive definer of available commands, retrievable by `NceTrafficController.getCmdGroups()`

Tautological sentences aside, we've got plenty of spare bits in `long cmdGroups` to define and centralise more capabilities there, e.g. `CMDS_PAGED_MODE`, `CMDS_DIRECT_MODE`,  `CMDS_REGISTER_MODE`... Would make maintenance and additions much easier. We've got a ridiculous number of capability selector definition lists to maintain in the NCE code.

dheap commented:
@bobjacobsen Can't check this today. All computers shut down since early this morning due to a day of lightning/thunder/hail/rain storms. No end in sight yet and tied up tonight.

dheap commented:
AppVeyor [2589](url
https://ci.appveyor.com/project/JMRI/jmri/builds/20634579) went very bad, restarted.

dheap commented:
What a nightmare of complexity!

Could not read with NCE DCC Twin.
See https://github.com/bobjacobsen/JMRI/pull/9

bobjacobsen commented:
Thanks! Have merged that update, and CI is running here.

I agree that this NCE code is really complicated. It would be good to centralize on e.g. `cmdGroups` as you suggest above.  But that might be an aspiration for another (set of) PR.

Is this one good enough to merge now so it's updated handling of internal & defaults can get some user feedback?

Thanks again

dheap commented:
Done NCE tests again. A couple of minor anomalies:

1. After running the Wizard to create an NCE USB connection (I think most), an RPS menu appears. It has gone next time you launch JMRI.
![screen shot 2018-11-30 at 1 33 05 pm](https://user-images.githubusercontent.com/13841457/49265643-12fbae80-f4a7-11e8-9d35-a2e87a33f9a3.png)
2. With the Power Pro via USB, The Simple Programmer shows "Write CV" as available, although it doesn't work. It does show "No programmer available!" in the status bar when you try.
![screen shot 2018-11-30 at 1 34 06 pm](https://user-images.githubusercontent.com/13841457/49266030-b00b1700-f4a8-11e8-9090-15558f116146.png)

I'd like to give SPROG another checkover...

@billchown offered to have a try with various SPROG hardware combinations against this branch and I'm hoping he will pick up on that.

dheap commented:
Tested creating SPROG and SPROG Command Station profiles using Wizard and all seems well:
![screen shot 2018-11-30 at 3 24 26 pm](https://user-images.githubusercontent.com/13841457/49268751-637a0880-f4b5-11e8-8f30-09b5b402b1ca.png)
![screen shot 2018-11-30 at 3 30 29 pm](https://user-images.githubusercontent.com/13841457/49268757-6ecd3400-f4b5-11e8-955b-501c07ed604a.png)

dheap commented:
Haven't rechecked migration of existing profiles from V4.12...

dheap commented:
Can't seem to reproduce the RPS error, even though I observed it a number of times during earlier testing.

Agree it may be time to merge this into master and get it out for others to test/comment...

bobjacobsen commented:
Re: "With the Power Pro via USB, The Simple Programmer shows "Write CV" as available, although it doesn't work. It does show "No programmer available!" in the status bar when you try."  Thanks for checking things like that and noticing issues.  That one's been around a long time, as the Simple Programmer doesn't seem to believe that a writing-capable programmer is ever not available.

I'll go ahead and merge.

billchown commented:
Dave,I've been out of town for a few days, but will get testing on the latest Jenkins on Tuesday.Thanks for all your tests, and bob for all that work - looks good in the ones that you showed.
Bill

Bill Chownorrrbrit2004@yahoo.comhttp://bchown.com/ 

    On Thursday, November 29, 2018, 7:09:12 PM PST, Dave Heap <notifications@github.com> wrote:  
 
 
Done NCE tests again. A couple of minor anomalies:
   
   - After running the Wizard to create an NCE USB connection (I think most), an RPS menu appears. It has gone next time you launch JMRI.   

   - With the Power Pro via USB, The Simple Programmer shows "Write CV" as available, although it doesn't work. It does show "No programmer available!" in the status bar when you try.   


I'd like to give SPROG another checkover...

@billchown offered to have a try with various SPROG hardware combinations against this branch and I'm hoping he will pick up on that.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.
